File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/square100/domain/Square100Game.java
package com.zwstudio.logicpuzzlesandroid.puzzles.square100.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;

import java.util.List;

import fj.F2;

/**
 * Created by zwvista on 2016/09/29.
 */

public class Square100Game extends CellsGame<Square100Game, Square100GameMove, Square100GameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };

    public String[] objArray;
    public String get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public String get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, String obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, String obj) {
        set(p.row, p.col, obj);
    }

    public Square100Game(List<String> layout, GameInterface<Square100Game, Square100GameMove, Square100GameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        size = new Position(layout.size(), layout.get(0).length());
        objArray = new String[rows() * cols()];

        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                String s = " " + str.charAt(c) + " ";
                set(r, c, s);
            }
        }

        Square100GameState state = new Square100GameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(Square100GameMove move, F2<Square100GameState, Square100GameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        Square100GameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(Square100GameMove move) {
        return changeObject(move, (state, move2) -> state.switchObject(move2));
    }

    public boolean setObject(Square100GameMove move) {
        return changeObject(move, (state, move2) -> state.setObject(move2));
    }

    public String getObject(Position p) {
        return state().get(p);
    }

    public String getObject(int row, int col) {
        return state().get(row, col);
    }

    public int getRowHint(int row) {
        return state().row2hint[row];
    }

    public int getColHint(int col) {
        return state().col2hint[col];
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/square100/Square100Game.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.square100

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class Square100Game(layout: List<String>, gi: GameInterface<Square100Game, Square100GameMove, Square100GameState>, gdi: GameDocumentInterface) : CellsGame<Square100Game, Square100GameMove, Square100GameState>(gi, gdi) {
    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1)
        )
    }

    var objArray: Array<String>
    operator fun get(row: Int, col: Int) = objArray[row * cols + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: String) {objArray[row * cols + col] = obj}
    operator fun set(p: Position, obj: String) {this[p.row, p.col] = obj}

    init {
        size = Position(layout.size, layout[0].length)
        objArray = Array(rows * cols) { "" }
        for (r in 0 until rows) {
            val str = layout[r]
            for (c in 0 until cols) {
                val s = " " + str[c] + " "
                this[r, c] = s
            }
        }
        val state = Square100GameState(this)
        levelInitilized(state)
    }

    fun getObject(p: Position) = currentState[p]
    fun getObject(row: Int, col: Int) = currentState[row, col]
    fun getRowHint(row: Int) = currentState.row2hint[row]
    fun getColHint(col: Int) = currentState.col2hint[col]
}